netutil/netutil.go

```diff
@@ -56,3 +59,166 @@ func privateNetworkInterfaces(all []net.Interface, fallback []string, logger log
 	level.Debug(logger).Log("msg", "found network interfaces with private IP addresses assigned", "interfaces", strings.Join(privInts, " "))
 	return privInts
 }
+
+// GetFirstAddressOf returns the first IPv4/IPV6 address of the supplied interface names, omitting any link-local addresses.
+func GetFirstAddressOf(names []string, logger log.Logger, preferInet6 bool) (string, error) {
+	return getFirstAddressOf(names, logger, getInterfaceAddresses, preferInet6)
+}
+
+// NetworkInterfaceGetter matches the signature of net.InterfaceByName() to allow for test mocks.
+type NetworkInterfaceAddressGetter func(name string) ([]netip.Addr, error)
+
+// getFirstAddressOf returns the first IPv4/IPV6 address of the supplied interface names, omitting any link-local addresses.
+func getFirstAddressOf(names []string, logger log.Logger, interfaceAddrsFunc NetworkInterfaceAddressGetter, preferInet6 bool) (string, error) {
+	var ipAddr netip.Addr
+
+	// When passing an empty list of interface names, we select all interfaces.
+	if len(names) == 0 {
+		infs, err := net.Interfaces()
+		if err != nil {
+			return "", fmt.Errorf("failed to get interface list and no interface names supplied: %w", err)
+		}
+		ifNames := make([]string, len(infs))
```

Comment:
[nit] Any reason why we don't directly use `names` instead?

Reply:
I was thinking that if we pass an empty list of `names`, then perhaps we haven't done the work to figure out what names are available, so in this block we do that work to determine all the names we've received.  I wasn't sure what else would be the right behavior if we receive an empty list.

Reply:
What I mean is: why defining a new variable `ifNames` instead of using the input `names` directly, so:

```
names := make([]string, len(infs))
for i, v := range infs {
  names[i] = v.Name
}
```

Reply:
Ah I see.  I'll adjust.